tsdb/db.go

```diff
@@ -1505,34 +1505,38 @@ func (db *DB) CleanTombstones() (err error) {
 	start := time.Now()
 	defer db.metrics.tombCleanTimer.Observe(time.Since(start).Seconds())
 
-	newUIDs := []ulid.ULID{}
-	defer func() {
-		// If any error is caused, we need to delete all the new directory created.
-		if err != nil {
-			for _, uid := range newUIDs {
-				dir := filepath.Join(db.Dir(), uid.String())
-				if err := os.RemoveAll(dir); err != nil {
-					level.Error(db.logger).Log("msg", "failed to delete block after failed `CleanTombstones`", "dir", dir, "err", err)
-				}
-			}
-		}
-	}()
+	// We can momentarily allow overlapping blocks in this particular case.
+	// This will allow the reload function to replace old blocks with the new ones.
+	// Notice that the the final result will be overlapping blocks, but with no tombstones.
+	if !db.opts.AllowOverlappingBlocks {
+		db.opts.AllowOverlappingBlocks = true
+		defer func() { db.opts.AllowOverlappingBlocks = false }()
+	}
 
-	db.mtx.RLock()
-	blocks := db.blocks[:]
-	db.mtx.RUnlock()
+	cleanUpCompleted := false
+	// Repeat cleanup until there is no tombstones left.
+	for !cleanUpCompleted {
+		cleanUpCompleted = true
 
-	for _, b := range blocks {
-		if uid, er := b.CleanTombstones(db.Dir(), db.compactor); er != nil {
-			err = errors.Wrapf(er, "clean tombstones: %s", b.Dir())
-			return err
-		} else if uid != nil { // New block was created.
-			newUIDs = append(newUIDs, *uid)
-		}
-	}
+		for _, pb := range db.Blocks() {
+			uid, cleanErr := pb.CleanTombstones(db.Dir(), db.compactor)
+			if cleanErr != nil {
+				return errors.Wrapf(cleanErr, "clean tombstones: %s", pb.Dir())
+			}
+			if uid == nil {
+				// There was nothing to clean
+				continue
+			}
 
-	if err := db.reloadBlocks(); err != nil {
-		return errors.Wrap(err, "reload blocks")
+			pb.meta.Compaction.Deletable = true
+			cleanUpCompleted = false
+			if err = db.reload(); err == nil { // Will try to delete old block.
+				// Successful reload will change the existing blocks.
+				// We need to loop over the new set of blocks.
+				break
+			}
+			return errors.Wrap(err, "reload blocks")
```

Comment:
We should be deleting the new block in case the reload fails.

Reply:
What I'm failing to understand is why adding this logic here fails the windows tests :disappointed:

Reply:
The new test shows that this problem doesn't occur on partial failures (some blocks get replaced and some dont) at least